### PR TITLE
CORE-19639 Rework of send/receive transaction builder flows

### PR DESCRIPTION
**Pull request by nkovacsx**
Main logic change:
Now we resolve the stateRefs to StateAndRefs to get the notary instead of making it mandatory all the time
Created a new payload for the builder that includes the builder data and an optional filtered dependencies:

- IF we have no dependencies at all
-> Only send the builder
 -> Receiver side sees that the builder has no dependencies -> No backchain resolution, no filtered dependency verification needed

- IF we have dependencies BUT no filtered dependencies
-> Only send the builder
-> Receiver side sees that we have dependencies BUT no filtered dependencies -> Need to perform backchain resolution

- IF we have dependencies AND filtered dependencies
-> Send both
-> Receiver side sees that we have BOTH dependencies and filtered dependencies -> Need to perform filtered transaction verification

### CODE REVIEW

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/ReceiveAndUpdateTransactionBuilderFlow.kt [hunk 1]
```diff
@@ -39,11 +40,14 @@ class ReceiveAndUpdateTransactionBuilderFlowVersionedFlowFactory(
 
     override fun create(version: Int, session: FlowSession): SubFlow<UtxoTransactionBuilder> {
         return when {
-            version >= 1 -> ReceiveAndUpdateTransactionBuilderFlowV1(
+            version > 1 -> ReceiveAndUpdateTransactionBuilderFlowV2(
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/ReceiveAndUpdateTransactionBuilderFlow.kt]:
```suggestion
            version >= CORDA_5_2.value -> ReceiveAndUpdateTransactionBuilderFlowV2(
```

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/SendTransactionBuilderDiffFlow.kt [hunk 2]
```diff
@@ -50,11 +51,14 @@ class SendTransactionBuilderDiffFlowVersionedFlowFactory(
 
     override fun create(version: Int, sessions: List<FlowSession>): SubFlow<Unit> {
         return when {
-            version >= 1 -> SendTransactionBuilderDiffFlowV1(
+            version > 1 -> SendTransactionBuilderDiffFlowV2(
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/SendTransactionBuilderDiffFlow.kt]: Same as above

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/v2/SendTransactionBuilderDiffFlowV2.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if (newTransactionIds.isEmpty()) {
+            log.trace {
+                "There are no new states transferred, therefore no backchain resolution or filtered dependencies required."
+            }
+            session.send(TransactionBuilderPayload(transactionBuilder))
+            return
+        }
+
+        // Resolve the dependent state and refs
+        val dependentStateAndRefs = ledgerStateQueryService.resolveStateRefs(
```
Comment (rick-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/v2/SendTransactionBuilderDiffFlowV2.kt]: This seems potentially expensive.  Couldn't we skip this if the notary is set?  Later on I presume something will fail during verify if the notary on the new transaction does not match.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/ReceiveAndUpdateTransactionBuilderFlow.kt (hunk 1 above)
Comment (rick-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/ReceiveAndUpdateTransactionBuilderFlow.kt]: I'm not convinced we need multiple versions yet.  We have agreed that 5.1 and 5.2 do not interop - and only ledger is inherited.


This just increases the number of permutations to test etc.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/v1/ReceiveAndUpdateTransactionBuilderFlowV1.kt [hunk 4]
```diff
@@ -98,7 +89,9 @@ class ReceiveAndUpdateTransactionBuilderFlowV1(
             }
 
             // Persist the verified filtered transactions
-            persistenceService.persistFilteredTransactionsAndSignatures(receivedFilteredTransactions)
+            receivedFilteredTransactions.let {
```
Comment (rick-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/transactionbuilder/v1/ReceiveAndUpdateTransactionBuilderFlowV1.kt]: This seems an odd way of writing it.  The original is more orthodox I think.

**Review by rick-r3 — DISMISSED** [review]
LGTM

**Review by rick-r3 — DISMISSED** [review]

**Review by rick-r3 — DISMISSED** [review]

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**